repository/copier.go

```
@@ -4,107 +4,191 @@ import (
+type Copier struct {
+	remote     Fs
+	local      *localFs
+	bucketSize int
 }
 
-// NewLocalCopier returns a Copier using as a remote a Billy filesystem
-func NewLocalCopier(fs billy.Filesystem, bucket int) Copier {
-	return &LocalCopier{fs, bucket}
+// NewCopier creates a new copier.
+func NewCopier(local billy.Filesystem, remote Fs, bucketSize int) *Copier {
+	return &Copier{remote: remote, local: &localFs{local}, bucketSize: bucketSize}
 }
 
-type LocalCopier struct {
-	fs         billy.Filesystem
-	bucketSize int
+// Local returns the local filesystem of the copier.
+func (c *Copier) Local() billy.Filesystem {
+	return c.local.Filesystem
 }
 
-func (c *LocalCopier) CopyFromRemote(src, dst string, localFs billy.Filesystem) error {
-	bSrc := addBucketName(src, c.bucketSize)
-	return c.copyFile(c.fs, localFs, bSrc, dst)
-}
+// CopyToRemote copies a file from the local filesystem to the remote one.
+func (c *Copier) CopyToRemote(ctx context.Context, src, dst string) error {
+	dstCopy := filepath.Join(c.remote.TempDir(), dst+".copy")
+	dst = filepath.Join(
+		c.remote.Base(),
+		addBucketName(dst, c.bucketSize),
+	)
 
-func (c *LocalCopier) CopyToRemote(src, dst string, localFs billy.Filesystem) error {
-	bDst := addBucketName(dst, c.bucketSize)
-	return c.copyFile(localFs, c.fs, src, bDst)
-}
+	lf, err := c.local.Open(filepath.Join(c.local.Base(), src))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
 
-func (c *LocalCopier) copyFile(fromFs, toFs billy.Filesystem, from, to string) (err error) {
-	src, err := fromFs.Open(from)
-	if os.IsNotExist(err) {
-		return nil
+	defer checkClose(lf, &err)
+
+	w, err := c.remote.WriteTo(dstCopy)
+	if err != nil {
+		return err
 	}
 
+	// Delete temporary file in case the process is stopped while copying
+	defer func() {
+		_ = w.Close()
+		_ = c.remote.DeleteIfExists(dstCopy)
+	}()
+
+	err = copy(ctx, w, lf)
+
+	checkClose(w, &err)
```

`checkClose` does it close when it's not closed? Because we already close the `w` in `defer func`.

---

That defer ignores all errors because it's only meant as a safety net in case the execution stopped for any reason and there's a panic. I could remove the defer, though. `w` is closed here and the file deleted later on, anyway.